changeloggenutils/merged_release.go

```diff
@@ -0,0 +1,252 @@
+    // will be included
+    MaxVersion *Version
+    // Minimum release version to display on this changelog. If not specified, all releases <= MaxVersion
+    // will be included
+    MinVersion *Version
+}
+
+type MergedReleaseGenerator struct {
+    client         *github.Client
+    releaseDepMap  map[Version]*Version
+    opts           Options
+    DependencyFunc DependencyFn
+}
+
+func NewMergedReleaseGenerator(opts Options, client *github.Client) *MergedReleaseGenerator {
+    generator :=  &MergedReleaseGenerator{
+        opts:          opts,
+        client:        client,
+        releaseDepMap: map[Version]*Version{},
+    }
+    generator.DependencyFunc = generator.GetOpenSourceDependency
+    return generator
+}
+
+func NewMergedReleaseGeneratorWithDepFn(opts Options, client *github.Client, depFn DependencyFn) *MergedReleaseGenerator {
+	gen :=  NewMergedReleaseGenerator(opts, client)
+	gen.DependencyFunc = depFn
+	return gen
+}
+
+/*
+ The merged release generator has 3 steps:
+ 1. Fetches enterprise repo release notes
+ 2. Gets open source dependency for each enterprise version
+ 3. Merges open source release notes into enterprise version release notes
+*/
+func (g *MergedReleaseGenerator) GenerateJSON(ctx context.Context) (string, error) {
+    var err error
+    enterpriseReleases, err := g.GetMergedEnterpriseRelease(ctx)
+    if err != nil {
+        return "", err
+    }
+    var out struct {
+        Opts        Options
+        ReleaseData *ReleaseData
+    }
+    out.Opts = g.opts
+    out.ReleaseData = enterpriseReleases
+    res, err := json.Marshal(out)
+    return string(res), err
+}
+
+func (g *MergedReleaseGenerator) GetMergedEnterpriseRelease(ctx context.Context) (*ReleaseData, error) {
+    ossReleases, err := NewMinorReleaseGroupedChangelogGenerator(Options{
+        RepoOwner: g.opts.RepoOwner,
+        MainRepo:  g.opts.DependentRepo,
+    }, g.client).
+        GetReleaseData(ctx)
+    if err != nil {
+        return nil, err
+    }
+    enterpriseReleases, err := NewMinorReleaseGroupedChangelogGenerator(g.opts, g.client).
+        GetReleaseData(ctx)
+    if err != nil {
+        return nil, err
+    }
+    return g.MergeEnterpriseReleaseWithOS(ctx, enterpriseReleases, ossReleases)
+}
+
+func (g *MergedReleaseGenerator) MergeEnterpriseReleaseWithOS(ctx context.Context, enterpriseReleases, osReleases *ReleaseData) (*ReleaseData, error) {
+    // Get openSourceReleases from max version to min version (e.g. 1.8.0, 1.8.0-beta2, 1.8.0-beta1...)
+    enterpriseSorted := enterpriseReleases.GetReleasesSorted()
+    osSorted := osReleases.GetReleasesSorted()
+    for _, release := range enterpriseSorted {
+        // Build out release dependency map (enterprise -> oss) using releaseDepMap as cache
+        if g.releaseDepMap[release] == nil {
+            var (
+                dep *Version
+                err error
+            )
+            dep, err = g.DependencyFunc(&release)
+            if err != nil {
```

Comment:
Should this error instead of continuing? This means that we were unable to find the dependency for a particular release. I assume this would only happen if (A) we had a bug in the function looking for the dep or (B) provided incorrect options for the main and dependent repos. Either way, it feels like we should fail loudly since something has been misconfigured by the user. What do you think?

Reply:
Since changelogs are generated at release time, I want to decrease the chance of a release failing due to user error. The worst case here is that open source dependencies don't show up for an enterprise releases' changelogs. I don't feel too strongly about it, but if a version tag is messed up for some reason, I'd rather have some other validation pick up that error than fail changelog generation completely.